### PR TITLE
Offline discussion updates for 8.0

### LINKED ISSUE #31688: Blazor 8.0 offline discussion updates
### Description

This is a general issue to track the updates from offline discussion for 8.0 content. I believe most (or all) of the updates can be batched on one PR. I'll open followup issues/PRs if needed for anything that can't be batched on a single PR.

* [x] Transient disposables ... **WORK SEPARATELY ON #31689**
* [x] BWA integration into the ASP.NET Core endpoint routing configuration. ... ***NO CHANGES NEEDED***
  * Optional params are covered: https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-8.0#route-parameters
  * Route constraints are covered (and optional params with constraints): https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-8.0#route-constraints
  * "Custom constraints" and "Regex constraints" were never added to the article per our goal to generally avoid covering features that Blazor doesn't offer. If it becomes a ❓ for readers and we receive feedback, I'll add something later.
* [x] `MapFallBackToPage`
* [x] "Don't place a <script> tag in a component file"
* [x] *Passing tokens* section ... **WORK SEPARATELY ON #31691**
* [x] *Cross-origin protection* section of the *Threat mitigation* topic ... **SKIPPING FOR NOW** ... we might add code examples if reader feedback indicates confusion.
* [x] *Configure and use gRPC in components* section ... **PARTIALLY ADDRESSED** by resurfacing the content sans any remarks about hosted WASM or any code directly tied to the server app of a hosted WASM solution. Dan said he can/will supply a code example.
* [x] *Secure a SignalR hub* section ... **PARTIALLY ADDRESSED** by resurfacing content sans any hosted WASM bits. Jeremy and/or Stephen will need to address the token question by DR.

### Page URL

Various

### Content source URL

Various

### Document ID

None

### Article author

@guardrex

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #31688
Addresses #28161 

## Items to be worked on separate issues

* ~[Transient disposable section updates](https://github.com/dotnet/AspNetCore.Docs/issues/31689): I have everything I need to resolve this, and I should reach it on Tuesday 🤞🍀.~ **UPDATE**: This is resolved now.
* [PU review of *Pass tokens to a server-side Blazor app* section](https://github.com/dotnet/AspNetCore.Docs/issues/31691): Currently assigned to both Stephen and Jeremy, but perhaps only one pass by one PU 🐱 will suffice.

## *Configure and use gRPC in components* section

Dan ... You said you'd provide a code example, but it 🤔 ***seems like*** the current code is ok. The updates to the section are on the DIFF, so you can see the current code there. The focus of the updates thus far is just to 🔪 OUT the little bits of text that pertain to hosted WASM and updating the versioning to return the content for 8.0+. Let me know if I should 🛑 HOLD this PR for new or additional code (or open a separate issue for it). 👂👂👂

## *Secure a SignalR hub* section

Dan asks Stephen/Jeremy ...

> Should we be recommending using the auth cookie instead? Are there any CSRF concerns here that would then need to be addressed?

The updates to the section are on the DIFF, so the current code is there for inspection. Thus far ... again ... the updates on the PR just focus on 🔪 OUT the little bits pertaining to hosted WASM and updating the versioning to return the content for 8.0+.

## Skipped bits

Everything else that we discussed that isn't on the PR doesn't require updates because the current coverage is good.

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/blazor/host-and-deploy/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/246e5c0d17394c1930b6bf511b10ca433b9607ff/aspnetcore/blazor/host-and-deploy/index.md) | [Host and deploy ASP.NET Core Blazor](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/index?branch=pr-en-us-31693) |
| [aspnetcore/blazor/javascript-interoperability/call-dotnet-from-javascript.md](https://github.com/dotnet/AspNetCore.Docs/blob/246e5c0d17394c1930b6bf511b10ca433b9607ff/aspnetcore/blazor/javascript-interoperability/call-dotnet-from-javascript.md) | [Call .NET methods from JavaScript functions in ASP.NET Core Blazor](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/call-dotnet-from-javascript?branch=pr-en-us-31693) |
| [aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md](https://github.com/dotnet/AspNetCore.Docs/blob/246e5c0d17394c1930b6bf511b10ca433b9607ff/aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md) | [Call JavaScript functions from .NET methods in ASP.NET Core Blazor](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/call-javascript-from-dotnet?branch=pr-en-us-31693) |
| [aspnetcore/blazor/javascript-interoperability/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/246e5c0d17394c1930b6bf511b10ca433b9607ff/aspnetcore/blazor/javascript-interoperability/index.md) | [ASP.NET Core Blazor JavaScript interoperability (JS interop)](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/index?branch=pr-en-us-31693) |
| [aspnetcore/blazor/security/webassembly/additional-scenarios.md](https://github.com/dotnet/AspNetCore.Docs/blob/246e5c0d17394c1930b6bf511b10ca433b9607ff/aspnetcore/blazor/security/webassembly/additional-scenarios.md) | [aspnetcore/blazor/security/webassembly/additional-scenarios](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/additional-scenarios?branch=pr-en-us-31693) |
| [aspnetcore/blazor/security/webassembly/index.md](https://github.com/dotnet/AspNetCore.Docs/blob/246e5c0d17394c1930b6bf511b10ca433b9607ff/aspnetcore/blazor/security/webassembly/index.md) | [Secure ASP.NET Core Blazor WebAssembly](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/security/webassembly/index?branch=pr-en-us-31693) |
| [aspnetcore/blazor/tutorials/signalr-blazor.md](https://github.com/dotnet/AspNetCore.Docs/blob/246e5c0d17394c1930b6bf511b10ca433b9607ff/aspnetcore/blazor/tutorials/signalr-blazor.md) | [[Visual Studio](#tab/visual-studio)](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/tutorials/signalr-blazor?branch=pr-en-us-31693) |

### CODE REVIEW

File: aspnetcore/blazor/javascript-interoperability/call-dotnet-from-javascript.md [hunk 1]
```diff
@@ -1422,11 +1422,20 @@ Using JS modules to load JS is described in this article in the [JavaScript isol
 
 :::moniker-end
 
-<!-- UPDATE 8.0 Confirm for >=8.0 now that the warning is being removed. -->
+:::moniker range=">= aspnetcore-8.0"
+
+> [!WARNING]
+> Only place a `<script>` tag in a component file (`.razor`) if the component is guaranteed to adopt [static server-side rendering (static SSR)](xref:blazor/fundamentals/index#client-and-server-rendering-concepts) because the `<script>` tag can't be updated dynamically.
```
Comment (halter73) [inline aspnetcore/blazor/javascript-interoperability/call-dotnet-from-javascript.md:1428]: @javiercn Is this warning worded strong enough for you? I remember you complaining about people trying to render `<script>` tags in triage.

**Review by halter73 — APPROVED** [review]

**Comment by guardrex** [comment]
Thanks, @halter73 ... but you didn't answer Dan's question in the opening remarks ☝️.

**Comment by halter73** [comment]
I think we should recommend using an auth cookie because that's the only way to prerender content that requires authentication.

As for the CSRF concerns, I think we should just link to https://learn.microsoft.com/en-us/aspnet/core/security/anti-request-forgery?view=aspnetcore-8.0. All Blazor endpoints that accept form posts that might be susceptible to CSRF attacks already require an antiforgery token as mentioned in https://learn.microsoft.com/en-us/aspnet/core/blazor/security/?view=aspnetcore-8.0#antiforgery-support, so I don't think there's much to add there. Pinging @javiercn @blowdart in case they disagree

Other types of endpoints that accept form posts either require antiforgery tokens by default (like razor pages, controllers with views, and minimal APIs) or have been long documented not to (like `AddControllers` without views). And the need for antiforgery checks for these other types of endpoints is unrelated to whether or not they are hosted along with a Blazor app.

**Comment by guardrex** [comment]
> I think we should recommend using an auth cookie because that's the only way to prerender content that requires authentication.

It will be a bit out-of-sync with our SignalR guidance ...

> Cookies are specific to browsers. Sending them from other kinds of clients adds complexity compared to sending bearer tokens. Cookie authentication isn't recommended unless the app only needs to authenticate users from the browser client. Bearer token authentication is the recommended approach when using clients other than the browser client.

Cross-ref: https://learn.microsoft.com/en-us/aspnet/core/signalr/authn-and-authz?view=aspnetcore-8.0#cookies-vs-bearer-tokens

We seem to have some cookie bits at ...

https://learn.microsoft.com/en-us/aspnet/signalr/overview/security/hub-authorization#cookie

@halter73 ... Do you want me to try to 🦖 ***RexHack***&trade; 🙈 some kind of approach based on that, or do we want a ***real programmer*** to build out what we'll show for the cookie approach in the SignalR doc?

BTW ... This is the current code that we show is, which I suppose we can maintain for standalone WASM apps ...

```csharp
@using Microsoft.AspNetCore.Components.WebAssembly.Authentication
@inject IAccessTokenProvider TokenProvider
@inject NavigationManager Navigation

...

var tokenResult = await TokenProvider.RequestAccessToken();

if (tokenResult.TryGetToken(out var token))
{
    hubConnection = new HubConnectionBuilder()
        .WithUrl(Navigation.ToAbsoluteUri("/chathub"), 
            options => { options.AccessTokenProvider = () => Task.FromResult(token?.Value); })
        .Build();

  ...
}
```

**Comment by guardrex** [comment]
## 🦖 NOTE TO SELF

After we finish the work for securing a SignalR hub, update the cross-links of the SignalR-Blazor tutorial, which currently only cross-link to the bearer token approach.

**Comment by halter73** [comment]
I think it's fine to continue recommend tokens for cases where there isn't prerendering like standalone WASM or non-browser apps. In those kind of apps, you don't have a browser window to login using Razor pages/components.

You can have a .NET HttpClient login using cookies with `MapIdentityApi` endpoints, but I think it's fair to say if you don't have a browser to do an OIDC flow or sign in using Razor pages/components, then bearer tokens are the better option.

If you're up for it, I'm more than happy to review 🦖 ***RexHack***&trade; version of cookie based SignalR auth using a .NET client. I expect instead of `options.AccessTokenProvider` you'd want to use `HttpMessageHandlerFactory` and configure it with `request.SetBrowserRequestCredentials(BrowserRequestCredentials.Include)` as demonstrated in https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/signalr?view=aspnetcore-8.0#client-side-signalr-cross-origin-negotiation-for-authentication

**Comment by guardrex** [comment]
@halter73 ... *Indeed!* I forgot that section was there. It seems to me that we can just cross-link that from the security article, as I just set up on ...

https://github.com/dotnet/AspNetCore.Docs/pull/31693/commits/fd91e23eac2345b159cb1e774352cf20d606d8c2

WRT ...

> Is this warning worded strong enough for you? I remember you complaining about people trying to render <script> tags in triage.

I already conversed with @javiercn on that point, and that's about what he told me to write into the docs.

WRT anything else ... @danroth27 / @mkArtakMSFT ... are you OK with going ahead and getting this merged?

File: aspnetcore/blazor/javascript-interoperability/call-dotnet-from-javascript.md (hunk 1 above)
Comment (danroth27) [reply aspnetcore/blazor/javascript-interoperability/call-dotnet-from-javascript.md:1428]: Seems fine to me. The reason for not using script tags in interactive components is clearly stated.

**Comment by danroth27** [comment]
> Dan ... You said you'd provide a code example, but it 🤔 seems like the current code is ok.

It looks like the gRPC sample code isn't used to initialize any component state, so yeah, I think it's fine as is. No need to persist and use prerendered state.

**Review by danroth27 — APPROVED** [review]
LGTM

**Pull request merged**